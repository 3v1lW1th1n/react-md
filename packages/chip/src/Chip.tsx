import React, { FC, HTMLAttributes, forwardRef, ReactNode } from "react";
import cn from "classnames";
import { TextIconSpacing } from "@react-md/icon";
import {
  InteractionStatesOptions,
  useInteractionStates,
} from "@react-md/states";
import { bem } from "@react-md/theme";
import { Omit, WithForwardedRef } from "@react-md/utils";

export interface ChipProps
  extends HTMLAttributes<HTMLButtonElement>,
    Omit<InteractionStatesOptions<HTMLButtonElement>, "disableSpacebarClick"> {
  /**
   * The button's type attribute. This is set to "button" by default so that forms are not
   * accidentally submitted when this prop is omitted since buttons without a type attribute work
   * as submit by default.
   */
  type?: "button" | "reset" | "submit";

  theme?: "outline" | "solid";

  leftIcon?: ReactNode;
  rightIcon?: ReactNode;
}

type WithRef = WithForwardedRef<HTMLButtonElement>;
type DefaultProps = Required<
  Pick<ChipProps, "theme" | "type" | "enablePressedAndRipple">
>;
type WithDefaultProps = ChipProps & DefaultProps & WithRef;

const block = bem("rmd-chip");

const Chip: FC<ChipProps & WithRef> = providedProps => {
  const {
    className: propClassName,
    children,
    forwardedRef,
    theme,
    leftIcon,
    rightIcon,
    disableRipple,
    disableProgrammaticRipple,
    rippleTimeout,
    rippleClassNames,
    rippleClassName,
    rippleContainerClassName,
    enablePressedAndRipple,
    ...props
  } = providedProps as WithDefaultProps;

  const { ripples, className, handlers } = useInteractionStates({
    handlers: props,
    className: propClassName,
    disabled: props.disabled,
    disableRipple,
    disableProgrammaticRipple,
    rippleTimeout,
    rippleClassNames,
    rippleClassName,
    rippleContainerClassName,
    enablePressedAndRipple,
  });

  return (
    <button
      {...props}
      {...handlers}
      ref={forwardedRef}
      className={cn(
        block({
          [theme]: true,
          "leading-icon": leftIcon,
          "trailing-icon": rightIcon,
        }),
        className
      )}
    >
      <TextIconSpacing icon={leftIcon}>
        <TextIconSpacing icon={rightIcon} iconAfter>
          {children}
        </TextIconSpacing>
      </TextIconSpacing>
      {ripples}
    </button>
  );
};

const defaultProps: DefaultProps = {
  theme: "solid",
  type: "button",
  enablePressedAndRipple: true,
};

Chip.defaultProps = defaultProps;

export default forwardRef<HTMLButtonElement, ChipProps>((props, ref) => (
  <Chip {...props} forwardedRef={ref} />
));
