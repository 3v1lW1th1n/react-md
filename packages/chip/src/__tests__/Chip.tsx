import React from "react";
import { render } from "@testing-library/react";

import Chip from "../Chip";

describe("Chip", () => {
  it("should render correctly", () => {
    const props = { children: "Content" };
    const { container, rerender } = render(<Chip {...props} />);

    expect(container).toMatchSnapshot();

    const leftIcon = <i className="material-icons">keyboard_arrow_left</i>;
    const rightIcon = <i className="material-icons">keyboard_arrow_right</i>;

    rerender(<Chip {...props} leftIcon={leftIcon} />);
    expect(container).toMatchSnapshot();

    rerender(<Chip {...props} rightIcon={rightIcon} />);
    expect(container).toMatchSnapshot();
  });
});
