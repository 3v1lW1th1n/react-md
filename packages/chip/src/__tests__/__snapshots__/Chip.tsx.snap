// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip should render correctly 1`] = `
<div>
  <button
    class="rmd-chip rmd-chip--solid"
    type="button"
  >
    Content
    <span
      class="rmd-ripple-container"
    />
  </button>
</div>
`;

exports[`Chip should render correctly 2`] = `
<div>
  <button
    class="rmd-chip rmd-chip--solid rmd-chip--leading-icon"
    type="button"
  >
    <i
      class="rmd-icon--before material-icons"
    >
      keyboard_arrow_left
    </i>
    Content
    <span
      class="rmd-ripple-container"
    />
  </button>
</div>
`;

exports[`Chip should render correctly 3`] = `
<div>
  <button
    class="rmd-chip rmd-chip--solid rmd-chip--trailing-icon"
    type="button"
  >
    Content
    <i
      class="rmd-icon--after material-icons"
    >
      keyboard_arrow_right
    </i>
    <span
      class="rmd-ripple-container"
    />
  </button>
</div>
`;
